### PR TITLE
feat: file-logger add vars

### PR DESCRIPTION
**Pull request by kamly**
### Description

file-loggerr  plugins add vars

fixes https://github.com/apache/apisix/issues/9596

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
@kamly please fix the code lint error

**Comment by monkeyDluffy6017** [comment]
test cases and docs are needed

**Comment by kamly** [comment]
hi, @monkeyDluffy6017 all done

- [ ] add test
- [ ] add doc
- [ ] fix code lint

File: t/plugin/file-logger2.t [hunk 1]
```diff
@@ -274,3 +274,95 @@ passed
...
+
+            if not fd then
+                msg = "not write file log"
+                ngx.say(msg)
+                return
+            end            
+        }
+    }
+--- response_body
+not write file log
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:368]: Need a blank

Comment (kamly) [reply t/plugin/file-logger2.t]: fix~

File: t/plugin/file-logger2.t [hunk 2]
```diff
@@ -274,3 +274,95 @@ passed
     }
 --- response_body
 write file log success
+
+=== TEST 8: Add new configuration with vars_match
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:278]: Three blanks are needed

File: t/plugin/file-logger2.t [hunk 3]
```diff
@@ -274,3 +274,95 @@ passed
...
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- response_body
+passed
+
+=== TEST 9: Request matches vars_match
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:317]: Ditto

File: t/plugin/file-logger2.t [hunk 4]
```diff
@@ -274,3 +274,95 @@ passed
...
+            end
+
+            os.remove("file-with-vars_match.log")
+        }
+    }
+--- response_body
+write file log success
+
+
+=== TEST 10: Request not matches vars_match
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:350]: Ditto

File: apisix/plugins/file-logger.lua [hunk 5]
```diff
@@ -113,6 +121,18 @@ if is_apisix_or then
     end
 end
 
+local function vars_match(vars, ctx)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/file-logger.lua:124]: Two blanks are needed here

File: apisix/plugins/file-logger.lua [hunk 6]
```diff
@@ -113,6 +121,18 @@ if is_apisix_or then
...
+        match_result = expr:eval(ctx.var)
+        if match_result then
+            break
+        end
+    end
+
+    return match_result
+end
 
 local function write_file_data(conf, log_message)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/file-logger.lua:137]: Ditto

File: docs/en/latest/plugins/file-logger.md [hunk 7]
```diff
@@ -49,6 +49,27 @@ The `file-logger` Plugin is used to push log streams to a specific location.
...
+        [ "arg_name","==","jack" ],
+        [ "arg_age","==",18 ]
+    ],
+    [
+        [ "arg_name2","==","allen" ]
+    ]
+]
+```
+
+以上示例表示前两个表达式之间的关系是 AND，而前两个和第三个表达式之间的关系是 OR。
```
Comment (monkeyDluffy6017) [inline docs/en/latest/plugins/file-logger.md:70]: Use english here

File: docs/zh/latest/plugins/file-logger.md [hunk 8]
```diff
@@ -51,6 +51,27 @@ description: API 网关 Apache APISIX file-logger 插件可用于将日志数据
...
+    ],
+    [
+        [ "arg_name2","==","allen" ]
+    ]
+]
+```
+
+以上示例表示前两个表达式之间的关系是 AND，而前两个和第三个表达式之间的关系是 OR。
+
+:::
```
Comment (monkeyDluffy6017) [inline docs/zh/latest/plugins/file-logger.md:74]: I don't think it's necessary to add the tip here, try to add an real example

File: t/plugin/file-logger2.t (hunk 3 above)
Comment (kamly) [reply t/plugin/file-logger2.t:317]: I'm sorry, what does ditto mean?

File: apisix/plugins/file-logger.lua [hunk 9]
```diff
@@ -38,6 +39,13 @@ local schema = {
             items = {
                 type = "array"
             }
+        },
+        vars = {
```
[the remark this answers is not in the document]
Comment (kamly) [reply apisix/plugins/file-logger.lua:43]: how about `matches` ?

File: docs/zh/latest/plugins/file-logger.md (hunk 8 above)
Comment (kamly) [reply docs/zh/latest/plugins/file-logger.md]: fix

File: docs/en/latest/plugins/file-logger.md (hunk 7 above)
Comment (kamly) [reply docs/en/latest/plugins/file-logger.md]: fix

File: apisix/plugins/file-logger.lua (hunk 6 above)
Comment (kamly) [reply apisix/plugins/file-logger.lua:139]: fix

File: apisix/plugins/file-logger.lua (hunk 5 above)
Comment (kamly) [reply apisix/plugins/file-logger.lua]: fix

File: t/plugin/file-logger2.t (hunk 3 above)
Comment (kamly) [reply t/plugin/file-logger2.t]: fix

File: apisix/plugins/file-logger.lua (hunk 9 above)
[the remark this answers is not in the document]
Comment (monkeyDluffy6017) [reply apisix/plugins/file-logger.lua]: ok

**Comment by kamly** [comment]
@monkeyDluffy6017 , fix again ~

File: docs/en/latest/plugins/file-logger.md [hunk 10]
```diff
@@ -49,6 +49,7 @@ The `file-logger` Plugin is used to push log streams to a specific location.
 | log_format | object | False    | Log format declared as key value pairs in JSON format. Values only support strings. [APISIX](../apisix-variable.md) or [Nginx](http://nginx.org/en/docs/varindex.html) variables can be used by prefixing the string with `$`. |
 | include_resp_body      | boolean | False     | When set to `true` includes the response body in the log file.                                                                                                                                                                |
 | include_resp_body_expr | array   | False     | When the `include_resp_body` attribute is set to `true`, use this to filter based on [lua-resty-expr](https://github.com/api7/lua-resty-expr). If present, only logs the response into file if the expression evaluates to `true`. |
+| matches        | array[] | False   |  Logs will be recorded only when the rule match is successful. See [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list) for a list of available expressions.   |
```
Comment (monkeyDluffy6017) [inline docs/en/latest/plugins/file-logger.md:52]:
```suggestion
| matches        | array[] | False   | Logs will be recorded when the rule matching is successful if the option is set. See [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list) for a list of available expressions.   |
```

File: docs/zh/latest/plugins/file-logger.md [hunk 11]
```diff
@@ -51,6 +51,7 @@ description: API 网关 Apache APISIX file-logger 插件可用于将日志数据
 | log_format       | object  | 否     | 以 JSON 格式的键值对来声明日志格式。对于值部分，仅支持字符串。如果是以 `$` 开头，则表明是要获取 [APISIX 变量](../apisix-variable.md) 或 [NGINX 内置变量](http://nginx.org/en/docs/varindex.html)。 |
 | include_resp_body      | boolean | 否     | 当设置为 `true` 时，生成的文件包含响应体。                                                                                               |
 | include_resp_body_expr | array   | 否     | 当 `include_resp_body` 属性设置为 `true` 时，使用该属性并基于 [lua-resty-expr](https://github.com/api7/lua-resty-expr) 进行过滤。如果存在，则仅在表达式计算结果为 `true` 时记录响应。       |
+| matches        | array[] | 否   |  当规则匹配通过后才会记录日志。`matches` 是一个表达式列表，来自 [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list)。   |
```
Comment (monkeyDluffy6017) [inline docs/zh/latest/plugins/file-logger.md:54]:
```suggestion
| matches        | array[] | 否   |  当设置了这个选项后，只有匹配规则的日志才会被记录。`matches` 是一个表达式列表，具体请参考 [lua-resty-expr](https://github.com/api7/lua-resty-expr#operator-list)。   |
```

File: docs/en/latest/plugins/file-logger.md [hunk 12]
```diff
@@ -128,3 +129,44 @@ curl http://127.0.0.1:9180/apisix/admin/routes/1  -H 'X-API-KEY: edd1c9f034335f1
   }
 }'
 ```
+
+## Configure the matches parameter
```
Comment (monkeyDluffy6017) [inline docs/en/latest/plugins/file-logger.md:133]: This block should be put before `Disable Plugin`

File: docs/zh/latest/plugins/file-logger.md [hunk 13]
```diff
@@ -143,3 +144,43 @@ curl http://127.0.0.1:9180/apisix/admin/routes/1  \
     }
 }'
 ```
+
+## 配置 matches 参数
```
Comment (monkeyDluffy6017) [inline docs/zh/latest/plugins/file-logger.md:148]: This block should be put before Section Disable Plugin

File: docs/en/latest/plugins/file-logger.md (hunk 12 above)
Comment (monkeyDluffy6017) [reply docs/en/latest/plugins/file-logger.md:133]: The section should be `Filter logs`

Comment (monkeyDluffy6017) [inline docs/en/latest/plugins/file-logger.md:133]:
```suggestion
## Filter logs
```

File: docs/zh/latest/plugins/file-logger.md (hunk 13 above)
Comment (monkeyDluffy6017) [inline docs/zh/latest/plugins/file-logger.md:148]:
```suggestion
## 过滤日志
```

Comment (kamly) [reply docs/zh/latest/plugins/file-logger.md]: fix~

File: docs/en/latest/plugins/file-logger.md (hunk 12 above)
Comment (kamly) [reply docs/en/latest/plugins/file-logger.md]: fix~

**Comment by kamly** [comment]
@monkeyDluffy6017 , fix again again ~

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by monkeyDluffy6017 — CHANGES_REQUESTED** [review]
please fix the ci error

**Comment by kamly** [comment]
@monkeyDluffy6017 fix fix ~

**Comment by Gallardot** [comment]
Refer to [include_req_body](https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/utils/log-util.lua#L166C13-L166C29) and `include_req_body_expr` to see if it is better to implement an `include_req_header` and `include_req_header_expr` feature so that other logging plugins can support this feature without each plugin implementing it separately.

@kamly @monkeyDluffy6017  WDYT?

**Comment by kamly** [comment]
> Refer to [include_req_body](https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/utils/log-util.lua#L166C13-L166C29) and `include_req_body_expr` to see if it is better to implement an `include_req_header` and `include_req_header_expr` feature so that other logging plugins can support this feature without each plugin implementing it separately.
> 
> @kamly @monkeyDluffy6017 WDYT?

good idea,  but i see file-logger not support `include_req_body_expr`

@monkeyDluffy6017 how do you think ?

**Comment by Gallardot** [comment]
> Good idea, but I read in the [docs](https://apisix.apache.org/zh/docs/apisix/plugins/file-logger/) that `file-logger` plugin doesn't support `include_req_body_expr`.

In fact, `file-logger` plugin support `include_req_body` and `include_req_body_expr`. You can refer to the [kafka plugin for configuration](https://apisix.apache.org/docs/apisix/plugins/kafka-logger/) and have a try.

https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/plugins/file-logger.lua#L151-L153
https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/utils/log-util.lua#L228
https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/utils/log-util.lua#L166

If it doesn't meet your scenario, you can implement the `include_req_header` and `include_req_header_expr` feature in `get_full_log`. 
https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/utils/log-util.lua#L108

**Comment by kamly** [comment]
> In fact, `file-logger` plugin support `include_req_body` and `include_req_body_expr`. You can refer to the [kafka plugin for configuration](https://apisix.apache.org/docs/apisix/plugins/kafka-logger/) and have a try.

I got it ~

But `file-logger` plugin does not support `include_req_body_expr` because the `schema` does not define `include_req_body_expr` . 

https://github.com/apache/apisix/blob/bb0a0e6e87eb0f15807be6fed0e7c983855d491c/apisix/plugins/file-logger.lua#L27-L44

Maybe the author who wrote it before missed it, I can try to align `kafka-logger` plugin

@monkeyDluffy6017 WDYT ?

**Comment by kamly** [comment]
@Gallardot Please take a look ?

**Comment by Gallardot** [comment]
> But `file-logger` plugin does not support `include_req_body_expr` because the `schema` does not define `include_req_body_expr` .

Although it is not defined in the schema, but it can be used directly in `file-logger` plugin.

**Comment by kamly** [comment]
> @kamly Although it is not defined in the schema, but it can be used directly in `file-logger` plugin. Therefore, I think it is only necessary to clearly define this feature in file-logger so that users can know more clearly that this feature is supported.

@Gallardot  got it, I have resubmitted following your solution.  Please take a look ?

in https://github.com/apache/apisix/pull/9712/commits/900e4386493992c318345406deefe182fa6af17a

**Comment by kamly** [comment]
@monkeyDluffy6017, Please take a look

**Comment by monkeyDluffy6017** [comment]
> so that other logging plugins can support this feature without each plugin implementing it separately.

Hi @Gallardot, I don't get your point, how can we do this?

**Comment by Gallardot** [comment]
> I don't get your point, how can we do this? each plugin is independent

Yes, each plugin is independent on runtime. However, in terms of implementation, some logging plug-ins use the same method `log_util.get_log_entry`.

**Comment by kamly** [comment]
@monkeyDluffy6017 What else do I need to update?

File: apisix/utils/log-util.lua [hunk 14]
```diff
@@ -267,6 +287,12 @@ function _M.check_log_schema(conf)
             return nil, "failed to validate the 'include_resp_body_expr' expression: " .. err
         end
     end
+    if conf.matches then
+        local ok, err = expr.new(conf.matches)
+        if not ok then
+            return nil, "failed to validate the 'matches' expression: " .. err
+        end
+    end
```
Comment (monkeyDluffy6017) [inline apisix/utils/log-util.lua:295]: No need to do this in the underlying function, you can do this in the plugin.
Not every plugin has this configuration `matches`

File: apisix/plugins/file-logger.lua [hunk 15]
```diff
@@ -38,6 +38,13 @@ local schema = {
             items = {
                 type = "array"
             }
+        },
+        matches = {
```
Comment (monkeyDluffy6017) [inline apisix/plugins/file-logger.lua:42]: I think `match` is better than `matches`

File: t/plugin/file-logger2.t [hunk 16]
```diff
@@ -274,3 +274,100 @@ passed
...
+            local code = t("/hello?name=jack", ngx.HTTP_GET)
+            local fd, err = io.open("file-with-matches.log", 'r')
+            local msg
+
+            if not fd then
+                core.log.error("failed to open file: file-with-matches.log, error info: ", err)
+                return
+            end
+
+            msg = fd:read()
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:336]:
```suggestion
            local fd, err = io.open("file-with-matches.log", 'r')
            if not fd then
                core.log.error("failed to open file: file-with-matches.log, error info: ", err)
                return
            end            
            local msg = fd:read()
```

File: t/plugin/file-logger2.t [hunk 17]
```diff
@@ -274,3 +274,100 @@ passed
...
+            local t = require("lib.test_admin").test
+            local code = t("/hello?name=tony", ngx.HTTP_GET)
+            local fd, err = io.open("file-with-matches.log", 'r')
+            local msg
+
+            if not fd then
+                msg = "not write file log"
+                ngx.say(msg)
+                return
+            end
```
Comment (monkeyDluffy6017) [inline t/plugin/file-logger2.t:369]:
```suggestion
            local code = t("/hello?name=tony", ngx.HTTP_GET)
            local fd, err = io.open("file-with-matches.log", 'r')
            if not fd then
                local msg = "not write file log"
                ngx.say(msg)
                return
            end
```

File: apisix/utils/log-util.lua (hunk 14 above)
Comment (kamly) [reply apisix/utils/log-util.lua:295]: @monkeyDluffy6017  
I think this judgment can be implemented in the underlying function. 
If other logger plugins need it, you only need to define `matches` in the `scheme` to have this feature

@Gallardot WDYT?

Comment (monkeyDluffy6017) [reply apisix/utils/log-util.lua:295]: But no other plugins need it now, we can move it here anytime if in need

File: apisix/plugins/file-logger.lua (hunk 15 above)
Comment (kamly) [reply apisix/plugins/file-logger.lua:42]: @monkeyDluffy6017 wouldn't it be better to use the plural form here?

File: apisix/utils/log-util.lua (hunk 14 above)
Comment (kamly) [reply apisix/utils/log-util.lua:295]: got it, i fix it

File: apisix/plugins/file-logger.lua (hunk 15 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/file-logger.lua:42]: I see the `match` is used everywhere in the APISIX, so just keep consistent

Comment (kamly) [reply apisix/plugins/file-logger.lua:42]: okok

**Comment by kamly** [comment]
@monkeyDluffy6017 all fix again ~

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by Gallardot — APPROVED** [review]

**Pull request merged**